plugins/outputs/postgresql/table_manager.go

```diff
@@ -0,0 +1,388 @@
+	var missingColumns []utils.Column
+	for _, srcCol := range srcColumns {
+		dbCol, ok := dbColumns[srcCol.Name]
+		if !ok {
+			missingColumns = append(missingColumns, srcCol)
+			continue
+		}
+		if !utils.PgTypeCanContain(dbCol.Type, srcCol.Type) {
+			return nil, fmt.Errorf("column type '%s' cannot store '%s'", dbCol.Type, srcCol.Type)
+		}
+	}
+	return missingColumns, nil
+}
+
+func (tm *TableManager) executeTemplates(
+	ctx context.Context,
+	db dbh,
+	tmpls []*template.Template,
+	tbl *tableState,
+	newColumns []utils.Column,
+	metricsTable *tableState,
+	tagsTable *tableState,
+) error {
+	tmplTable := template.NewTable(tm.Schema, tbl.name, colMapToSlice(tbl.Columns()))
+	metricsTmplTable := template.NewTable(tm.Schema, metricsTable.name, colMapToSlice(metricsTable.Columns()))
+	var tagsTmplTable *template.Table
+	if tagsTable != nil {
+		tagsTmplTable = template.NewTable(tm.Schema, tagsTable.name, colMapToSlice(tagsTable.Columns()))
+	} else {
+		tagsTmplTable = template.NewTable("", "", nil)
+	}
+
+	/* https://github.com/jackc/pgx/issues/872
+	stmts := make([]string, len(tmpls))
+	batch := &pgx.Batch{}
+	for i, tmpl := range tmpls {
+		sql, err := tmpl.Render(tmplTable, newColumns, metricsTmplTable, tagsTmplTable)
+		if err != nil {
+			return err
+		}
+		stmts[i] = string(sql)
+		batch.Queue(stmts[i])
+	}
+
+	batch.Queue(refreshTableStructureStatement, tm.Schema, tableName)
+
+	batchResult := tm.db.SendBatch(ctx, batch)
+	defer batchResult.Close()
+
+	for i := 0; i < len(tmpls); i++ {
+		if x, err := batchResult.Exec(); err != nil {
+			return fmt.Errorf("executing `%.40s...`: %v %w", stmts[i], x, err)
+		}
+	}
+
+	rows, err := batchResult.Query()
+	if err != nil {
+		return fmt.Errorf("refreshing table: %w", err)
+	}
+	tm.refreshTableStructureResponse(tableName, rows)
+	*/
+
+	// Lock to prevent concurrency issues in postgres (pg_type_typname_nsp_index unique constraint; SQLSTATE 23505)
+	tm.schemaMutex.Lock()
+	defer tm.schemaMutex.Unlock()
+
+	tx, err := db.Begin(ctx)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback(ctx)
+
+	for _, tmpl := range tmpls {
+		sql, err := tmpl.Render(tmplTable, newColumns, metricsTmplTable, tagsTmplTable)
+		if err != nil {
+			return err
+		}
+		if _, err := tx.Exec(ctx, string(sql)); err != nil {
+			return fmt.Errorf("executing `%s`: %w", sql, err)
+		}
+	}
+
+	// We need to be able to determine the role of the column when reading the structure back (because of the templates).
+	// For some columns we can determine this by the column name (time, tag_id, etc). However tags and fields can have any
+	// name, and look the same. So we add a comment to tag columns, and through process of elimination what remains are
+	// field columns.
+	for _, col := range newColumns {
+		if col.Role != utils.TagColType {
+			continue
+		}
+		if _, err := tx.Exec(ctx, "COMMENT ON COLUMN "+tmplTable.String()+"."+template.QuoteIdentifier(col.Name)+" IS 'tag'"); err != nil {
```

Comment:
Constructing the query with an `fmt.Sprintf` in a separate line is probably easier to read.